### PR TITLE
add Detection output op for SSD

### LINKED ISSUE #6225: Detection Output Operator.
for ssd Detection Output

### PR DESCRIPTION
**Pull request by sweetsky0901**
fix  #6225

### CODE REVIEW

File: paddle/operators/detection_output_op.cc [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#include "paddle/operators/detection_output_op.h"
+namespace paddle {
+namespace operators {
+
+class Detection_output_OpMaker : public framework::OpProtoAndCheckerMaker {
```
[the remark this answers is not in the document]
Comment (sweetsky0901) [reply paddle/operators/detection_output_op.cc]: done

File: paddle/operators/detection_output_op.cc [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+ public:
+  Detection_output_OpMaker(framework::OpProto* proto,
```
[the remark this answers is not in the document]
Comment (sweetsky0901) [reply paddle/operators/detection_output_op.cc]: done

File: paddle/operators/detection_output_op.h [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+    auto dst_stride = framework::stride(dst->dims());
+    auto src_stride = framework::stride(in_p_tensor_transpose.dims());
+    StridedMemcpy<T>(context.device_context(), in_p_tensor_transpose.data<T>(),
+                     src_stride, in_p_tensor_transpose.dims(), dst_stride,
+                     dst->data<T>() + offset);
+    offset += in_p_tensor_transpose.dims()[4] * src_stride[4];
+  }
+}
+template <typename DeviceContext, typename T>
+class Detection_output_Kernel : public framework::OpKernel<T> {
```
Comment (wanghaox) [inline paddle/operators/detection_output_op.h]: class name use camel-case naming.

File: paddle/operators/math/detection_util.h [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+  T y_min;
+  T x_max;
+  T y_max;
+  // whether difficult object (e.g. object with heavy occlusion is difficult)
+  bool is_difficult;
+};
+// KNCHW ==> NHWC
+// template <typename T>
+template <typename T>
+void getBBoxFromPriorData(const T* prior_data, const size_t num_bboxes,
```
Comment (wanghaox) [inline paddle/operators/math/detection_util.h]: use camel-case naming, the first letter needs to be capitalized.

File: paddle/operators/math/detection_util.h [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+                          std::vector<BBox<T>>& bbox_vec);
+template <typename T>
+void getBBoxVarFromPriorData(const T* prior_data, const size_t num,
+                             std::vector<std::vector<T>>& var_vec);
+template <typename T>
+BBox<T> decodeBBoxWithVar(BBox<T>& prior_bbox,
```
Comment (wanghaox) [inline paddle/operators/math/detection_util.h]: use camel-case naming, the first letter needs to be capitalized.

File: paddle/operators/math/detection_util.h (hunk 4 above)
Comment (sweetsky0901) [reply paddle/operators/math/detection_util.h]: done

File: paddle/operators/math/detection_util.h (hunk 5 above)
Comment (sweetsky0901) [reply paddle/operators/math/detection_util.h]: done

File: paddle/operators/detection_output_op.h (hunk 3 above)
Comment (sweetsky0901) [reply paddle/operators/detection_output_op.h]: done

File: paddle/operators/CMakeLists.txt [hunk 6]
```diff
@@ -187,6 +186,36 @@ endfunction()
 add_subdirectory(math)
 add_subdirectory(nccl)
```
Comment (luotao1) [inline paddle/operators/CMakeLists.txt]: 这个PR merge的有点早。这里cmake的冲突来自于 https://github.com/PaddlePaddle/Paddle/pull/7067 ，但不应该直接加189-218行。

**Comment by wangkuiyi** [comment]
https://github.com/PaddlePaddle/Paddle/blob/43dab723c8a2757e5736df53f68cdb880d67b9f6/paddle/operators/detection_output_op.h#L53-L164 -- a single function with more than 100 lines of code -- @wanghaox how could you approve and merge such an extremely low-quality PR?

**Pull request merged**